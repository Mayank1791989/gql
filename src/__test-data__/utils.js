/* @flow */
import path from 'path';
import { dedent } from 'dentist';
import splitLines from '../shared/splitLines';

export function getSchema() {
  jest.mock('../shared/watch');
  const watch = require('../shared/watch').default;
  const GQLService = require('../').GQLService;

  const gql = new GQLService({ cwd: path.resolve('src/__test-data__/') });
  // $FlowDisableNextLine
  watch.__triggerChange();
  return gql._schemaBuilder.getSchema();
}

export function getDefLocations() {
  return {
    Player: {
      start: { line: 12, column: 1 },
      end: { line: 16, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Player_id: {
      start: { line: 13, column: 3 },
      end: { line: 13, column: 10 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Player_image_arg_size: {
      start: { line: 15, column: 9 },
      end: { line: 15, column: 19 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Team: {
      start: { line: 17, column: 1 },
      end: { line: 20, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Node: {
      start: { line: 8, column: 1 },
      end: { line: 10, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Edge: {
      start: { line: 27, column: 1 },
      end: { line: 30, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Mutation: {
      start: { line: 2, column: 1 },
      end: { line: 5, column: 2 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreate: {
      start: { line: 4, column: 3 },
      end: { line: 4, column: 63 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreateInput: {
      start: { line: 6, column: 1 },
      end: { line: 9, column: 2 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },
    Mutation_PlayerCreateInput_id: {
      start: { line: 8, column: 3 },
      end: { line: 8, column: 10 },
      path: path.resolve('src/__test-data__/schema/mutation.gql'),
    },

    Query: {
      start: { line: 2, column: 1 },
      end: { line: 6, column: 2 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },
    Query_viewer: {
      start: { line: 5, column: 3 },
      end: { line: 5, column: 18 },
      path: path.resolve('src/__test-data__/schema/query.gql'),
    },

    Role: {
      start: { line: 1, column: 1 },
      end: { line: 5, column: 2 },
      path: path.resolve('src/__test-data__/schema/enums.gql'),
    },

    CustomScalar: {
      start: { line: 1, column: 1 },
      end: { line: 1, column: 20 },
      path: path.resolve('src/__test-data__/schema/scalars.gql'),
    },

    customDirective: {
      start: { line: 2, column: 1 },
      end: { line: 7, column: 25 },
      path: path.resolve('src/__test-data__/schema/directives.gql'),
    },
    customDirective_argIf: {
      start: { line: 4, column: 3 },
      end: { line: 4, column: 15 },
      path: path.resolve('src/__test-data__/schema/directives.gql'),
    },
  };
}

export function getRefLocations() {
  return {
    Player: [
      {
        start: { line: 14, column: 11 },
        end: { line: 14, column: 17 },
        path: path.resolve('src/__test-data__/schema/mutation.gql'),
      },
      {
        start: { line: 24, column: 7 },
        end: { line: 24, column: 13 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 12, column: 6 },
        end: { line: 12, column: 12 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
    ],

    Edge: [
      {
        start: { line: 28, column: 11 },
        end: { line: 28, column: 15 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
    ],

    Role: [
      {
        start: { line: 10, column: 9 },
        end: { line: 10, column: 13 },
        path: path.resolve('src/__test-data__/schema/mutation.gql'),
      },
      {
        start: { line: 34, column: 9 },
        end: { line: 34, column: 13 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 36, column: 15 },
        end: { line: 36, column: 19 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 1, column: 6 },
        end: { line: 1, column: 10 },
        path: path.resolve('src/__test-data__/schema/enums.gql'),
      },
    ],

    CustomScalar: [
      {
        start: { line: 35, column: 10 },
        end: { line: 35, column: 22 },
        path: path.resolve('src/__test-data__/schema/query.gql'),
      },
      {
        start: { line: 1, column: 8 },
        end: { line: 1, column: 20 },
        path: path.resolve('src/__test-data__/schema/scalars.gql'),
      },
    ],
  };
}

export function getHints() {
  return {
    ObjectTypes: [
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
    ],

    OutputTypes: [
      {
        description: '',
        text: 'Role',
        type: 'Enum',
      },
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
      {
        description: '',
        text: 'CustomScalar',
        type: 'Scalar',
      },
      {
        description: 'The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.',
        text: 'String',
        type: 'Scalar',
      },
      {
        description: 'The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ',
        text: 'Int',
        type: 'Scalar',
      },
      {
        description: 'The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ',
        text: 'Float',
        type: 'Scalar',
      },
      {
        description: 'The `Boolean` scalar type represents `true` or `false`.',
        text: 'Boolean',
        type: 'Scalar',
      },
      {
        description: 'The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `"4"`) or integer (such as `4`) input value will be accepted as an ID.',
        text: 'ID',
        type: 'Scalar',
      },
    ],

    CompositeTypes: [
      {
        description: 'Mutation contains all allowed mutations',
        text: 'Mutation',
        type: 'Object',
      },
      {
        description: '',
        text: 'PlayerCreatePayload',
        type: 'Object',
      },
      {
        description: 'Query is the root query object',
        text: 'Query',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Viewer',
        type: 'Object',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
      {
        description: '',
        text: 'NewPlayer',
        type: 'Object',
      },
    ],

    InputTypes: [
      {
        description: '',
        text: 'Role',
        type: 'Enum',
      },
      {
        description: '',
        text: 'PlayerCreateInput',
        type: 'Input',
      },
      {
        description: '',
        text: 'CustomScalar',
        type: 'Scalar',
      },
      {
        description: 'The `String` scalar type represents textual data, represented as UTF-8 character sequences. The String type is most often used by GraphQL to represent free-form human-readable text.',
        text: 'String',
        type: 'Scalar',
      },
      {
        description: 'The `Int` scalar type represents non-fractional signed whole numeric values. Int can represent values between -(2^31) and 2^31 - 1. ',
        text: 'Int',
        type: 'Scalar',
      },
      {
        description: 'The `Float` scalar type represents signed double-precision fractional values as specified by [IEEE 754](http://en.wikipedia.org/wiki/IEEE_floating_point). ',
        text: 'Float',
        type: 'Scalar',
      },
      {
        description: 'The `Boolean` scalar type represents `true` or `false`.',
        text: 'Boolean',
        type: 'Scalar',
      },
      {
        description: 'The `ID` scalar type represents a unique identifier, often used to refetch an object or as key for a cache. The ID type appears in a JSON response as a String; however, it is not intended to be human-readable. When expected as an input type, any string (such as `"4"`) or integer (such as `4`) input value will be accepted as an ID.',
        text: 'ID',
        type: 'Scalar',
      },
    ],

    InterfaceTypes: [
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
      {
        description: '',
        text: 'Edge',
        type: 'Interface',
      },
    ],

    PlayerTypeFields: [
      {
        description: '',
        text: 'id',
        type: 'ID!',
      },
      {
        description: '',
        text: 'name',
        type: 'String!',
      },
      {
        description: '',
        text: 'image',
        type: 'String!',
      },
    ],

    PlayerTypeImageFieldArgs: [
      {
        description: '',
        text: 'size',
        type: 'Int!',
      },
    ],

    TypesImplementsNode: [
      {
        description: '',
        text: 'Player',
        type: 'Object',
      },
      {
        description: '',
        text: 'Team',
        type: 'Object',
      },
      {
        description: '',
        text: 'Node',
        type: 'Interface',
      },
    ],

    PossibleMutations: [
      {
        description: 'Player Create contains all allowed mutations',
        text: 'PlayerCreate',
        type: 'PlayerCreatePayload',
      },
    ],

    QueryFields: [
      {
        description: '',
        text: 'node',
        type: 'Node',
      },
      {
        description: '',
        text: 'nodes',
        type: '[Node]',
      },
      {
        description: '',
        text: 'viewer',
        type: 'Viewer!',
      },
      {
        description: 'Access the current type schema of this server.',
        text: '__schema',
        type: '__Schema!',
      },
      {
        description: 'Request the type information of a single type.',
        text: '__type',
        type: '__Type',
      },
    ],

    EnumRoleValues: [
      {
        description: '',
        text: 'roleA',
        type: 'Role',
      },
      {
        description: '',
        text: 'roleB',
        type: 'Role',
      },
      {
        description: '',
        text: 'roleC',
        type: 'Role',
      },
    ],
  };
}

export function code(text: string) {
  const sourceText = dedent(text);
  const lines = splitLines(sourceText);
  let position = null;
  lines.forEach((line, index) => {
    const match = line.match(/--\^/);
    if (match) {
      position = {
        // $FlowDisableNextLine
        column: match.index + 3,
        line: index,
      };
    }
  });

  if (!position) {
    throw new Error('Missing --^ in source');
  }

  return {
    sourceText,
    position,
  };
}
