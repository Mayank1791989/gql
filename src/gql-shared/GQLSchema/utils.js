/* @flow */
import {
  type TypeNode,
  type NamedTypeNode,
  type EnumValueDefinitionNode,
  type FieldDefinitionNode,
  getDirectiveValues,
  GraphQLDeprecatedDirective,
  Kind,
} from 'graphql';

export function getNamedTypeNode(typeNode: TypeNode): NamedTypeNode {
  let namedType = typeNode;
  while (
    namedType.kind === Kind.LIST_TYPE ||
    namedType.kind === Kind.NON_NULL_TYPE
  ) {
    namedType = namedType.type;
  }
  return namedType;
}

/**
 * Given a field or enum value node, returns the string value for the
 * deprecation reason.
 */
export function getDeprecationReason(
  node: EnumValueDefinitionNode | FieldDefinitionNode,
): ?string {
  const deprecated = getDirectiveValues(GraphQLDeprecatedDirective, node);
  return deprecated && (deprecated.reason: any);
}
