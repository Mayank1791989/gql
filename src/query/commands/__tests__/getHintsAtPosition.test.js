/* @flow */
import { getHintsAtPosition } from '../index';
import { getHints, getSchema, code } from '../../../__test-data__/utils';

const hints = getHints();
const relayConfig = {
  parser: ['EmbeddedQueryParser', { startTag: 'Relay.QL`', endTag: '`' }],
  isRelay: true,
};
const queryConfig = {
  parser: 'QueryParser',
};

const schema = getSchema();

describe('fragments', () => {
  it('on Type suggestions', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on   #
          ---------------^
      \`;
    `);

    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.CompositeTypes);
  });

  it('on Type suggestions after few characters', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Pl  #
          ----------------^
      \`;
    `);

    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.CompositeTypes);
  });

  it('fields suggestions: Simple', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
                #
         #---^
        }
      \`;
    `);

    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.PlayerTypeFields);
  });

  it('alias fields', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          some:    #
          ------^
        }
      \`;
    `);

    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.PlayerTypeFields);
  });

  it('[Relay] fields after interpolation', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          id
          \${Component.getFragment('player')}
                 #
          #---^
        }
      \`;
    `);

    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.PlayerTypeFields);
  });

  it('fields args', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Player {
          image()
          ------^
        }
      \`;
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.PlayerTypeImageFieldArgs);
  });

  it('inline fragments on Type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Node {
          id
          ...on   #
          ------^
        }
      \`;
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.TypesImplementsNode);
  });

  it('inline fragments on Type fields', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment test on Node {
          id
          ...on Player {
                #
           #--^
          }
        }
      \`;
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.PlayerTypeFields);
  });
});

describe('mutation', () => {
  it('fields', () => {
    const { sourceText, position } = code(`
      Relay.QL\`
        mutation { }
        ----------^
      \`;
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.PossibleMutations);
  });

  it('field args', () => {
    const { sourceText, position } = code(`
      mutation {
        PlayerCreate()
        -------------^
      }
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, queryConfig),
    ).toMatchSnapshot();
  });
});

describe('Values', () => {
  it('enum value suggesstion', () => {
    const { sourceText, position } = code(`
      fragment test on NewPlayer {
        image(role: )
        #----------^
      }
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, queryConfig),
    ).toEqual(hints.EnumRoleValues);
  });

  it('Object value keys', () => {
    const { sourceText, position } = code(`
      mutation {
        PlayerCreate(
          input: {
            id: "string",
              #
        #---^
          }
        })
      }
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, queryConfig),
    ).toMatchSnapshot();
  });
});

describe('query', () => {
  it('fields', () => {
    const { sourceText, position } = code(`
      Relay.QL\`
        query Test { }
        ------------^
      \`;
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toEqual(hints.QueryFields);
  });
});

describe('directives', () => {
  it('directives', () => {
    const { sourceText, position } = code(`
      fragment on Test {
        id @inc
      --------^
      }
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, queryConfig),
    ).toMatchSnapshot();
  });

  it('directives: include relay directives', () => {
    const { sourceText, position } = code(`
      Relay.QL\`
        fragment on Test @relay {
                #----------^
        }
      \`;
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toMatchSnapshot();
  });

  it('args', () => {
    const { sourceText, position } = code(`
      Relay.QL\`
        fragment on Test @relay() {
            --------------------^
          id
        }
      \`;
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toMatchSnapshot();
  });

  it('args: value', () => {
    const { sourceText, position } = code(`
      Relay.QL\`
        fragment on Test @relay(pattern:  ) {
                                ---------^
          id
        }
      \`;
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toMatchSnapshot();
  });
});


describe('show meta field __typename in abstract types', () => {
  it('interface type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment on Node {
          id
          __type
        #-------^
        }
      \`
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toMatchSnapshot();
  });

  it('union type', () => {
    const { sourceText, position } = code(`
      const a = Relay.QL\`
        fragment on Entity {
          __type
        #-------^
        }
      \`
    `);
    expect(
      getHintsAtPosition(schema, sourceText, position, relayConfig),
    ).toMatchSnapshot();
  });
});
