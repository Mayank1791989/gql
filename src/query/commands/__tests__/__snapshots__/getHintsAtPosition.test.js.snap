// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Values Object value keys 1`] = `
Array [
  Object {
    "description": "",
    "text": "id",
    "type": "ID!",
  },
  Object {
    "description": "",
    "text": "name",
    "type": "String!",
  },
  Object {
    "description": "",
    "text": "role",
    "type": "Role!",
  },
]
`;

exports[`directives args 1`] = `
Array [
  Object {
    "description": "Marks a connection field as containing nodes without \`id\` fields. This is used to silence the warning when diffing connections.",
    "text": "isConnectionWithoutNodeID",
    "type": "Boolean",
  },
  Object {
    "description": "Marks a fragment as static. A static fragment will share the same identity regardless of how many times the expression is evaluated.",
    "text": "isStaticFragment",
    "type": "Boolean",
  },
  Object {
    "description": "Marks a fragment as intended for pattern matching (as opposed to fetching).",
    "text": "pattern",
    "type": "Boolean",
  },
  Object {
    "description": "Marks a fragment as being backed by a GraphQLList",
    "text": "plural",
    "type": "Boolean",
  },
  Object {
    "description": "Selectively pass variables down into a fragment.",
    "text": "variables",
    "type": "[String]",
  },
]
`;

exports[`directives args: value 1`] = `
Array [
  Object {
    "description": "Not false.",
    "text": "true",
    "type": "Boolean",
  },
  Object {
    "description": "Not true.",
    "text": "false",
    "type": "Boolean",
  },
]
`;

exports[`directives directives 1`] = `
Array [
  Object {
    "description": "some custom directive",
    "text": "customDirective",
  },
  Object {
    "description": "Directs the executor to skip this field or fragment when the \`if\` argument is true.",
    "text": "skip",
  },
  Object {
    "description": "Directs the executor to include this field or fragment only when the \`if\` argument is true.",
    "text": "include",
  },
]
`;

exports[`directives directives: include relay directives 1`] = `
Array [
  Object {
    "description": "some custom directive",
    "text": "customDirective",
  },
  Object {
    "description": "The @relay directive.",
    "text": "relay",
  },
]
`;

exports[`mutation field args 1`] = `
Array [
  Object {
    "description": "",
    "text": "input",
    "type": "PlayerCreateInput!",
  },
]
`;

exports[`show meta field __typename in abstract types interface type 1`] = `
Array [
  Object {
    "description": "",
    "text": "id",
    "type": "ID!",
  },
  Object {
    "description": "The name of the current Object type at runtime.",
    "text": "__typename",
    "type": "String!",
  },
]
`;

exports[`show meta field __typename in abstract types union type 1`] = `
Array [
  Object {
    "description": "The name of the current Object type at runtime.",
    "text": "__typename",
    "type": "String!",
  },
]
`;
